chalice/package.py

```diff
@@ -594,3 +605,67 @@ def _fixup_deployment_package(self, template, outdir):
                 self._osutils.copy(original_location, new_location)
                 copied = True
             resource['Properties']['CodeUri'] = './deployment.zip'
+
+
+class TemplateMergePostProcessor(TemplatePostProcessor):
+    def process(self, template, config, outdir, chalice_stage_name):
+        # type: (Dict[str, Any], Config, str, str) -> None
+        if not self._should_merge_template(config):
+            return
+        loaded_template = self._load_template_to_merge(config)
+        merger = TemplateDeepMerger()
+        merged = merger.merge(loaded_template, template)
+        template.clear()
+        template.update(merged)
+
+    def _should_merge_template(self, config):
+        # type: (Config) -> bool
+        return config.package_merge_template is not None
+
+    def _load_template_to_merge(self, config):
+        # type: (Config) -> Dict[str, Any]
+        filepath = os.path.abspath(config.package_merge_template)
+        if not self._osutils.file_exists(filepath):
+            raise RuntimeError('Cannot find template file: %s' % filepath)
+        template_data = self._osutils.get_file_contents(filepath, binary=False)
+        try:
+            loaded_template = json.loads(template_data)
+        except ValueError:
+            raise RuntimeError(
+                'Expected %s to be valid JSON template.' % filepath)
+        return loaded_template
+
+
+class TemplateDeepMerger(object):
+    def merge(self, src, dst):
+        # type: (Dict[str, Any], Dict[str, Any]) -> Dict[str, Any]
+        return self._merge_dict(src, dst)
```

Comment:
It was a little hard to follow that we skip over the general purpose `_merge` and go right to the `_merge_dict`.  I don't know if I'm over simplifying this but, ignoring the behavior of deleting `None`s and also keeping the existing `src` wins behavior (which I think we should change as per my comment above), I would have expected this is as simple as:

```
    def merge(self, src, dst):
        return self._merge(src, dst)

    def _merge(self, src, dst):
        if isinstance(src, dict) and isinstance(dst, dict):
            return self._merge_dict(src, dst)
        return src

    def _merge_dict(self, src, dst):
        merged = dst.copy()
        for key, value in src.items():
            merged[key] = self._merge(value, dst.get(key))
        return merged

```

Caveat: haven't actually checked if this works.

I think if we start implementing different deep merge strategies then we can look at breaking this down into a template method.

Reply:
Yea so I had much more complex merging logic initially and I wound up removing a lot of it the more I thought about it, so some of this is because it was being used and I was leaving it there in case we wanted to add more cases back like having special logic for lists etc. But if everyone is happy with the basic starting merge behavior it has now, simplifying implementation is good with me.